File: go/internal/ui/avatar.go
package ui

import (
	"fmt"
	"strings"
)

// AvatarSize represents the possible sizes for an avatar
type AvatarSize int

const (
	// AvatarSize16 represents a 16px avatar
	AvatarSize16 AvatarSize = 16
	// AvatarSize24 represents a 24px avatar
	AvatarSize24 AvatarSize = 24
	// AvatarSize32 represents a 32px avatar
	AvatarSize32 AvatarSize = 32
	// AvatarSize48 represents a 48px avatar
	AvatarSize48 AvatarSize = 48
	// AvatarSize64 represents a 64px avatar
	AvatarSize64 AvatarSize = 64
	// AvatarSize96 represents a 96px avatar
	AvatarSize96 AvatarSize = 96
)

// AvatarItem represents an item that can be displayed as an avatar
type AvatarItem interface {
	GetName() string
	GetAvatarURL() string
	GetID() int
}

// AvatarEmail represents an email address that can be displayed as an avatar
type AvatarEmail struct {
	Email string
}

// GetName returns the email address as the name
func (e AvatarEmail) GetName() string {
	return e.Email
}

// GetAvatarURL returns an empty string as there is no avatar URL for an email
func (e AvatarEmail) GetAvatarURL() string {
	return ""
}

// GetID returns 0 as there is no ID for an email
func (e AvatarEmail) GetID() int {
	return 0
}

// Avatar represents an avatar component
type Avatar struct {
	Item          AvatarItem
	Alt           string
	CustomClass   string
	Size          AvatarSize
	AvatarOptions map[string]interface{}
}

// NewAvatar creates a new avatar with the given parameters
func NewAvatar(item AvatarItem, alt string, customClass string, size AvatarSize, avatarOptions map[string]interface{}) *Avatar {
	if avatarOptions == nil {
		avatarOptions = make(map[string]interface{})
	}

	return &Avatar{
		Item:          item,
		Alt:           alt,
		CustomClass:   customClass,
		Size:          size,
		AvatarOptions: avatarOptions,
	}
}

// AvatarClasses returns the CSS classes for the avatar
func (a *Avatar) AvatarClasses() string {
	var classes []string

	// Add base classes
	classes = append(classes, "gl-avatar", fmt.Sprintf("gl-avatar-s%d", a.Size), a.CustomClass)

	// Add shape class based on item type
	switch a.Item.(type) {
	case AvatarEmail:
		classes = append(classes, "gl-avatar-circle")
	default:
		classes = append(classes, "!gl-rounded-base")
	}

	// Add identicon classes if no source
	if a.Src() == "" {
		classes = append(classes, "gl-avatar-identicon")
		classes = append(classes, fmt.Sprintf("gl-avatar-identicon-bg%d", (a.Item.GetID()%7)+1))
	}

	return strings.Join(classes, " ")
}

// Src returns the source URL for the avatar
func (a *Avatar) Src() string {
	if a.Item == nil {
		return ""
	}

	// If the item has an avatar URL, use it
	if url := a.Item.GetAvatarURL(); url != "" {
		return fmt.Sprintf("%s?width=%d", url, a.Size)
	}

	return ""
}

// Srcset returns the srcset attribute for the avatar
func (a *Avatar) Srcset() string {
	src := a.Src()
	if src == "" {
		return ""
	}

	retinaSrc := strings.Replace(src, fmt.Sprintf("width=%d", a.Size), fmt.Sprintf("width=%d", a.Size*2), 1)
	return fmt.Sprintf("%s 1x, %s 2x", src, retinaSrc)
}

// Alt returns the alt text for the avatar
func (a *Avatar) AltText() string {
	if a.Alt != "" {
		return a.Alt
	}
	return a.Item.GetName()
}

// Initial returns the first letter of the item's name
func (a *Avatar) Initial() string {
	name := a.Item.GetName()
	if name == "" {
		return ""
	}
	return strings.ToUpper(string(name[0]))
}

// Render generates the HTML for the avatar
func (a *Avatar) Render() string {
	// Build avatar attributes
	avatarAttrs := make(map[string]interface{})
	for k, v := range a.AvatarOptions {
		avatarAttrs[k] = v
	}
	avatarAttrs["class"] = a.AvatarClasses()
	avatarAttrs["alt"] = a.AltText()

	// If we have a source, render an image
	if src := a.Src(); src != "" {
		avatarAttrs["src"] = src
		avatarAttrs["srcset"] = a.Srcset()
		avatarAttrs["height"] = a.Size
		avatarAttrs["width"] = a.Size
		avatarAttrs["loading"] = "lazy"

		return fmt.Sprintf(`<img %s>`, formatAttributes(avatarAttrs))
	}

	// Otherwise, render a div with the initial
	return fmt.Sprintf(`<div %s>%s</div>`, formatAttributes(avatarAttrs), a.Initial())
}

File: rust_app/src/components/pajamas/avatar.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

/// Avatar item types
#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum AvatarItem {
    /// User avatar
    User {
        id: i64,
        name: String,
        email: String,
        avatar_url: Option<String>,
        blocked: bool,
    },
    /// Project avatar
    Project {
        id: i64,
        name: String,
        avatar_url: Option<String>,
    },
    /// Group avatar
    Group {
        id: i64,
        name: String,
        avatar_url: Option<String>,
    },
    /// Email avatar
    Email(String),
    /// Direct URL avatar
    Url(String),
}

impl AvatarItem {
    fn name(&self) -> String {
        match self {
            AvatarItem::User { name, .. } => name.clone(),
            AvatarItem::Project { name, .. } => name.clone(),
            AvatarItem::Group { name, .. } => name.clone(),
            AvatarItem::Email(email) => email.clone(),
            AvatarItem::Url(url) => url.clone(),
        }
    }

    fn id(&self) -> Option<i64> {
        match self {
            AvatarItem::User { id, .. } => Some(*id),
            AvatarItem::Project { id, .. } => Some(*id),
            AvatarItem::Group { id, .. } => Some(*id),
            _ => None,
        }
    }

    fn is_user_or_email(&self) -> bool {
        matches!(self, AvatarItem::User { .. } | AvatarItem::Email(_))
    }

    fn avatar_url(&self, size: i32) -> Option<String> {
        match self {
            AvatarItem::User {
                avatar_url, email, ..
            } => avatar_url.clone().or_else(|| {
                Some(format!(
                    "https://secure.gravatar.com/avatar/{}?s={}&d=identicon",
                    md5::compute(email.as_bytes()),
                    size
                ))
            }),
            AvatarItem::Project { avatar_url, .. } | AvatarItem::Group { avatar_url, .. } => {
                avatar_url
                    .clone()
                    .map(|url| format!("{}?width={}", url, size))
            }
            AvatarItem::Email(email) => Some(format!(
                "https://secure.gravatar.com/avatar/{}?s={}&d=identicon",
                md5::compute(email.as_bytes()),
                size
            )),
            AvatarItem::Url(url) => Some(url.clone()),
        }
    }
}

/// Represents an email for avatar generation
#[derive(Debug, Clone)]
pub struct AvatarEmail {
    pub email: String,
}

impl AvatarEmail {
    pub fn new(email: String) -> Self {
        Self { email }
    }

    pub fn name(&self) -> String {
        self.email.clone()
    }
}

/// Size options for the avatar component
pub const SIZE_OPTIONS: [i32; 6] = [16, 24, 32, 48, 64, 96];

/// Renders a Pajamas compliant avatar element
#[derive(Debug, Clone)]
pub struct AvatarComponent {
    /// The item to display (User, Project, Group, AvatarEmail, or String URL)
    item: Box<dyn std::any::Any + Send + Sync>,
    /// Text for the alt attribute
    alt: Option<String>,
    /// Custom CSS class(es)
    class: String,
    /// Size in pixels
    size: i32,
    /// Additional avatar options
    avatar_options: HashMap<String, String>,
}

impl AvatarComponent {
    /// Create a new avatar component
    pub fn new<T: 'static + Send + Sync>(
        item: T,
        alt: Option<String>,
        class: String,
        size: i32,
        avatar_options: Option<HashMap<String, String>>,
    ) -> Self {
        Self {
            item: Box::new(item),
            alt,
            class,
            size: Self::filter_size(size),
            avatar_options: avatar_options.unwrap_or_default(),
        }
    }

    /// Filter size to ensure it's a valid option
    fn filter_size(size: i32) -> i32 {
        if SIZE_OPTIONS.contains(&size) {
            size
        } else {
            64 // default size
        }
    }

    /// Get the avatar classes
    fn avatar_classes(&self) -> String {
        let mut classes = vec![
            "gl-avatar".to_string(),
            format!("gl-avatar-s{}", self.size),
            self.class.clone(),
        ];

        // Add circle class for User or AvatarEmail
        if self.is_user_or_email() {
            classes.push("gl-avatar-circle".to_string());
        } else {
            classes.push("!gl-rounded-base".to_string());
        }

        // Add identicon classes if no src
        if self.src().is_none() {
            classes.push("gl-avatar-identicon".to_string());
            let id = self.get_id().unwrap_or(0);
            classes.push(format!("gl-avatar-identicon-bg{}", (id % 7) + 1));
        }

        classes.join(" ")
    }

    /// Check if item is User or AvatarEmail
    fn is_user_or_email(&self) -> bool {
        self.item.is::<User>() || self.item.is::<AvatarEmail>()
    }

    /// Get the item ID
    fn get_id(&self) -> Option<i64> {
        // This would need to be implemented based on your User/Project/Group types
        None
    }

    /// Get the source URL for the avatar
    fn src(&self) -> Option<String> {
        if let Some(url) = self.item.downcast_ref::<String>() {
            return Some(url.clone());
        }

        if let Some(user) = self.item.downcast_ref::<User>() {
            // This would need to be implemented based on your User type
            return None;
        }

        if let Some(email) = self.item.downcast_ref::<AvatarEmail>() {
            // This would need to be implemented based on your email avatar generation
            return None;
        }

        // Try to get avatar_url from the item
        if let Some(url) = self.get_avatar_url() {
            return Some(format!("{}?width={}", url, self.size));
        }

        None
    }

    /// Get the srcset for retina displays
    fn srcset(&self) -> Option<String> {
        self.src().map(|src| {
            let retina_src = src.replace(
                &format!("width={}", self.size),
                &format!("width={}", self.size * 2),
            );
            format!("{} 1x, {} 2x", src, retina_src)
        })
    }

    /// Get the alt text
    fn alt(&self) -> String {
        self.alt
            .clone()
            .unwrap_or_else(|| self.get_name().unwrap_or_default())
    }

    /// Get the initial for identicon
    fn initial(&self) -> String {
        self.get_name()
            .map(|name| {
                name.chars()
                    .next()
                    .unwrap_or(' ')
                    .to_uppercase()
                    .to_string()
            })
            .unwrap_or_default()
    }

    /// Get the name from the item
    fn get_name(&self) -> Option<String> {
        // This would need to be implemented based on your types
        None
    }

    /// Get the avatar URL from the item
    fn get_avatar_url(&self) -> Option<String> {
        // This would need to be implemented based on your types
        None
    }

    /// Render the avatar component
    pub fn render(&self) -> String {
        if let Some(src) = self.src() {
            let mut attrs = vec![
                format!("src=\"{}\"", src),
                format!("alt=\"{}\"", self.alt()),
                format!("class=\"{}\"", self.avatar_classes()),
                format!("height=\"{}\"", self.size),
                format!("width=\"{}\"", self.size),
                "loading=\"lazy\"".to_string(),
            ];

            if let Some(srcset) = self.srcset() {
                attrs.push(format!("srcset=\"{}\"", srcset));
            }

            // Add additional avatar options
            for (key, value) in &self.avatar_options {
                attrs.push(format!("{}=\"{}\"", key, value));
            }

            format!("<img {}>", attrs.join(" "))
        } else {
            let mut attrs = vec![
                format!("alt=\"{}\"", self.alt()),
                format!("class=\"{}\"", self.avatar_classes()),
            ];

            // Add additional avatar options
            for (key, value) in &self.avatar_options {
                attrs.push(format!("{}=\"{}\"", key, value));
            }

            format!("<div {}>{}</div>", attrs.join(" "), self.initial())
        }
    }
}

// This is a placeholder for the User type - you'll need to implement your own User type
#[derive(Debug, Clone)]
pub struct User {
    // Add necessary fields
}

// Helper trait for type checking
trait TypeCheck {
    fn is<T: 'static>(&self) -> bool;
}

impl TypeCheck for Box<dyn std::any::Any + Send + Sync> {
    fn is<T: 'static>(&self) -> bool {
        self.as_any().is::<T>()
    }
}

// Helper trait for downcasting
trait AsAny {
    fn as_any(&self) -> &dyn std::any::Any;
}

impl AsAny for Box<dyn std::any::Any + Send + Sync> {
    fn as_any(&self) -> &dyn std::any::Any {
        self.as_ref()
    }
}
